blockchain/message_pool/src/msgpool.rs

```diff
@@ -0,0 +1,1066 @@
+    bls_sig_cache: Arc<RwLock<LruCache<Cid, Signature>>>,
+    msg: UnsignedMessage,
+) -> Result<SignedMessage, Error> {
+    let mut bls_sig_cache = bls_sig_cache.write().await;
+    let val = bls_sig_cache
+        .get(&msg.cid()?)
+        .ok_or_else(|| Error::Other("Could not recover sig".to_owned()))?;
+    Ok(SignedMessage::new_from_parts(msg, val.clone()))
+}
+
+/// Finish verifying signed message before adding it to the pending mset hashmap. If an entry
+/// in the hashmap does not yet exist, create a new mset that will correspond to the from message
+/// and push it to the pending hashmap
+async fn add_helper<T>(
+    api: Arc<RwLock<T>>,
+    bls_sig_cache: Arc<RwLock<LruCache<Cid, Signature>>>,
+    pending: Arc<RwLock<HashMap<Address, MsgSet>>>,
+    msg: SignedMessage,
+) -> Result<(), Error>
+where
+    T: Provider,
+{
+    let mut pending = pending.write().await;
+    let api = api.read().await;
+    let mut bls_sig_cache = bls_sig_cache.write().await;
+    if msg.signature().signature_type() == SignatureType::BLS {
+        bls_sig_cache.put(msg.cid()?, msg.signature().clone());
+    }
+
+    if msg.message().gas_limit() > 100_000_000 {
+        return Err(Error::Other(
+            "given message has too high of a gas limit".to_string(),
+        ));
+    }
+
+    api.put_message(&msg)?;
+    let msett = pending.get_mut(msg.message().from());
+    match msett {
+        Some(mset) => mset.add(&msg)?,
+        None => {
+            let mut mset = MsgSet::new();
+            mset.add(&msg)?;
+            pending.insert(*msg.message().from(), mset);
+        }
+    }
+    Ok(())
+}
+
+/// This function will revert and/or apply tipsets to the message pool. This function should be
+/// called every time that there is a head change in the message pool
+pub async fn head_change<T>(
+    api: Arc<RwLock<T>>,
+    bls_sig_cache: Arc<RwLock<LruCache<Cid, Signature>>>,
+    pending: Arc<RwLock<HashMap<Address, MsgSet>>>,
+    cur_tipset: Arc<RwLock<Tipset>>,
+    revert: Vec<Tipset>,
+    apply: Vec<Tipset>,
+) -> Result<(), Error>
+where
+    T: Provider,
+{
+    let mut rmsgs: HashMap<Address, HashMap<u64, SignedMessage>> = HashMap::new();
+    for ts in revert {
+        let api_locked = api.write().await;
+        let pts = api_locked.load_tipset(ts.parents())?;
+        drop(api_locked);
+
+        let mut msgs: Vec<SignedMessage> = Vec::new();
+        for block in ts.blocks() {
+            let api_locked = api.read().await;
+            let (umsg, mut smsgs) = api_locked.messages_for_block(&block)?;
+            drop(api_locked);
+            msgs.append(smsgs.as_mut());
+            for msg in umsg {
+                let bls_sig_cache = bls_sig_cache.clone();
+                let smsg = recover_sig(bls_sig_cache, msg).await?;
+                msgs.push(smsg)
+            }
+        }
+        let parent = pts.clone();
+        let mut cur_ts_locked = cur_tipset.write().await;
+        *cur_ts_locked = parent;
```

Comment:
```suggestion
        let mut cur_ts_locked = cur_tipset.write().await;
        *cur_ts_locked = pts;
```
that is a very expensive clone btw